src/NuGet.Services.AzureSearch/DocumentUtilities.cs

```diff
@@ -0,0 +1,239 @@
+            document.Prerelease = leaf.IsPrerelease;
+            document.ProjectUrl = leaf.ProjectUrl;
+            document.Published = leaf.Published;
+            document.ReleaseNotes = leaf.ReleaseNotes;
+            document.RequiresLicenseAcceptance = leaf.RequireLicenseAgreement;
+            document.SemVerLevel = leaf.IsSemVer2() ? 2 : (int?)null;
+            document.Summary = leaf.Summary;
+            document.Tags = leaf.Tags == null ? null : leaf.Tags.ToArray();
+            document.Title = leaf.Title;
+        }
+
+        public static string GetSearchDocumentKey(string packageId, SearchFilters searchFilters)
+        {
+            var lowerId = packageId.ToLowerInvariant();
+            var encodedId = EncodeKey(lowerId);
+            return $"{encodedId}-{searchFilters}";
+        }
+
+        public static string GetHijackDocumentKey(string packageId, string normalizedVersion)
+        {
+            var lowerId = packageId.ToLowerInvariant();
+            var lowerVersion = normalizedVersion.ToLowerInvariant();
+            return EncodeKey($"{lowerId}/{lowerVersion}");
+        }
+
+        private static string EncodeKey(string rawKey)
+        {
+            // First, encode the raw value for uniqueness.
+            var bytes = Encoding.UTF8.GetBytes(rawKey);
+            var unique = HttpServerUtility.UrlTokenEncode(bytes);
+
+            // Then, prepend a string as close to the raw key as possible, for readability.
+            var readable = ReplaceUnsafeKeyCharacters(rawKey).TrimStart('_');
+
+            return readable.Length > 0 ? $"{readable}-{unique}" : unique;
+        }
+
+        private static string ReplaceUnsafeKeyCharacters(string input)
+        {
+            return Regex.Replace(
+                input,
+                "[^A-Za-z0-9-_]", // Remove equal sign as well, since it's ugly.
+                "_",
+                RegexOptions.None,
+                TimeSpan.FromSeconds(30));
+        }
+
+        private static DateTimeOffset? AssumeUtc(DateTime? dateTime)
+        {
+            if (!dateTime.HasValue)
+            {
+                return null;
+            }
+
+            return new DateTimeOffset(dateTime.Value.Ticks, TimeSpan.Zero);
+        }
+
+        /// <summary>
+        /// This method produces output for official client. The implementation on the client side, at one point of time
+        /// was:
+        /// https://github.com/NuGet/NuGet.Client/blob/b404acf6eb88c2b6086a9cbb5106104534de2428/src/NuGet.Core/NuGet.Protocol/LegacyFeed/V2FeedPackageInfo.cs#L228-L308
+        /// 
+        /// The output is a string where each dependency is seperated by a "|" character and information about each
+        /// dependency is seperated by a ":" character. Per dependency, the colon-seperated data is a pair or triple.
+        /// The first field is the dependency package ID. If it is an empty string, then the version range must also be
```

Comment:
"[I]t" in the second sentence is ambiguous.  How about:

    The three fields in order are:  dependency package ID, version range, and target framework.  If the first field is an empty string, ...

Reply:
Sure!